core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java

```diff
@@ -0,0 +1,421 @@
+            this.index = index;
+            this.shardId = shardId;
+            this.reason = reason;
+            status = ExceptionsHelper.status(reason);
+        }
+
+        private Failure() {
+        }
+
+        public String nodeId() {
+            return nodeId;
+        }
+
+        @Override
+        public String index() {
+            return index;
+        }
+
+        @Override
+        public int shardId() {
+            return shardId;
+        }
+
+        @Override
+        public String reason() {
+            return detailedMessage(reason);
+        }
+
+        @Override
+        public RestStatus status() {
+            return status;
+        }
+
+        @Override
+        public Throwable getCause() {
+            return reason;
+        }
+
+        public static Failure readFailure(StreamInput in) throws IOException {
+            Failure failure = new Failure();
+            failure.readFrom(in);
+            return failure;
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            nodeId = in.readString();
+            index = in.readString();
+            shardId = in.readInt();
+            reason = in.readThrowable();
+            status = RestStatus.readFrom(in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(nodeId);
+            out.writeString(index);
+            out.writeInt(shardId);
+            out.writeThrowable(reason);
+            RestStatus.writeTo(out, status);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field("node", nodeId());
+            builder.field("shard", shardId());
+            builder.field("index", index());
+            builder.field("status", status.name());
+            if (reason != null) {
+                builder.field("reason");
+                builder.startObject();
+                ElasticsearchException.toXContent(builder, params, reason);
+                builder.endObject();
+            }
+            return builder;
+        }
+    }
+
+    private ImmutableOpenMap<String, ImmutableOpenIntMap<List<ShardStoreStatus>>> shardStatuses;
+    private ImmutableList<Failure> failures;
+
+    IndicesShardsStoresResponse(ImmutableOpenMap<String, ImmutableOpenIntMap<List<ShardStoreStatus>>> shardStatuses, ImmutableList<Failure> failures) {
+        this.shardStatuses = shardStatuses;
+        this.failures = failures;
+    }
+
+    IndicesShardsStoresResponse() {
+        this(ImmutableOpenMap.<String, ImmutableOpenIntMap<List<ShardStoreStatus>>>of(), ImmutableList.<Failure>of());
+    }
+
+    /**
+     * Returns {@link ShardStoreStatus}s
+     * grouped by their index names and shard ids.
+     */
+    public ImmutableOpenMap<String, ImmutableOpenIntMap<List<ShardStoreStatus>>> getShardStatuses() {
```

Comment:
makes sense, I removed the Shard prefix for most of the classes/method names in `IndicesShardsStoresResponse `